### DISCUSSION TITLE
Airflow REST API: Support getting information about cross-dag dependencies via GET request

### DISCUSSION THREAD

**Discussion by tmccall8829** [apache/airflow#30885] [General]
### Description

Hey all! I'm working on building an internal tool for visualizing the dependency graph and gantt chart for all of our cross-dag dependencies in our entire airflow instance. I was planning on building the graph on top of data retrieved from the Airflow v2 REST API, specifically the `/dags` and the `/dags/{dag_id}/tasks` endpoints.

I was hoping to be able to see in the API response for TriggerDagRunOperator tasks some info about the DAG that task triggers, but I don't see anything in the response for DAGs or DAG tasks that I can use for this purpose. I only see within-DAG task dependencies, nothing about cross-dag dependencies. Is there an endpoint in the API (or, alternatively, maybe something stored in the metadata db) I could use for that purpose?

It seems like maybe the way the API serves up DAG and task info is very confined to within-DAG information and not cross-dag info... but please correct me if I'm wrong or misunderstanding how the API should be used 🙂

### Use case/motivation

For example, let's say I have a really simple setup with two DAGs: A and B. A runs first, and at the end of the run has a TriggerDagRunOperator that triggers DAG B by name. So the simple dependency graph there would be something like:
`A_t1 <- A_t2 <- (B_t*)` (i.e., DAG B and all associated tasks only run when triggered by A_t2).

I was thinking that it'd be nice, e.g., to be able to GET `/dags/B` and have it include something like:
```
{
    ...
    "dag_id": "B",
    "upstream_dag_trigger": ["A"],
    ...
}
```
(Or, alternatively, request info about A_t2 and have it list something like "triggered_downstream_dag" name).

I guess theoretically I could construct this myself by reading in A's source code via REST API and read the actual task info for the TriggerDagRunOperator task as defined in the DAG itself, but that feels cumbersome for an airflow setup with, say, hundreds of DAGs. Do you all have any ideas there? Am I sort of just asking for something that can't really be done?

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
I think what you are really looking at are Dataset support. TriggerDagRun is just an operator and we are not planning to add more semantics meaning and make specific API for that. From the point of view this is  specific operator "trigger DAGrun" that has speicific parameters (dag_id to trigger) and if you want to parse it and build your dependencies there, you are free to do so.

But really the way we added (and we plan to build on) when it comes to dag dependencies is dataset-driven schedule and dataset dependencies and it has not only capabilities of keeping relations between dags and tasks, but also add semantic meaning to the dataset produced and consumed by those, and it will be acompanied with Lineage information as well.  So you almost definiely want to use datasets for what you want to do https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_dataset

Converting it to a discussion - this is of course my personal view, but it is highly unlikely we are going to enhance generic DAG/TASK API with more semantics information about the dependencies, where Datasets are the future.

Converting it to a discussion if more of it is needed.

**Comment by Macr0Nerd** [comment]
I will say that while I agree that dataset driven dependencies are a better option, the dependency information is already collected and shown in the DAG Dependency Graph, and having that information available through the API would be very helpful. My company, for instance, is migrating to Airflow so it's going to be a bit before we're able to utilize all its capabilities, including dataset driven workflows. But this dependency information is still necessary so we have to build out custom tooling for data that's already provided on the UI, but not API. Anyways, thanks for the great work!
